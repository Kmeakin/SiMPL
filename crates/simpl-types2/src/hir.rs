pub use simpl_syntax2::ast::{Ident, Lit};
use simpl_syntax2::{ast, parse, ParseError};

pub type ExprId = u32;

#[derive(Debug, Copy, Clone, Default)]
struct IdGen {
    counter: u32,
}

impl IdGen {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn next(&mut self) -> u32 {
        let x = self.counter;
        self.counter += 1;
        x
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Lit {
        id: ExprId,
        val: Lit,
    },
    Var {
        id: ExprId,
        name: String,
    },
    If {
        id: ExprId,
        test: Box<Expr>,
        then_branch: Box<Expr>,
        else_branch: Box<Expr>,
    },
    Let {
        id: ExprId,
        bindings: Vec<(Ident, Expr)>,
        body: Box<Expr>,
    },
    Letrec {
        id: ExprId,
        bindings: Vec<(Ident, Expr)>,
        body: Box<Expr>,
    },
    Lambda {
        id: ExprId,
        params: Vec<Ident>,
        body: Box<Expr>,
    },
    App {
        id: ExprId,
        func: Box<Expr>,
        arg: Box<Expr>,
    },
}

impl Expr {
    pub fn from_ast(ast: ast::Expr) -> Expr {
        let mut gen = IdGen::new();

        Expr::from_ast_inner(ast, &mut gen)
    }

    pub fn from_str(src: &str) -> Result<Expr, ParseError> {
        let ast = parse(src)?;
        Ok(Expr::from_ast(ast))
    }

    fn from_ast_inner(ast: ast::Expr, gen: &mut IdGen) -> Expr {
        let id = gen.next();

        match ast {
            ast::Expr::Lit { val } => Expr::Lit { id, val },
            ast::Expr::Var { name } => Expr::Var { id, name },
            ast::Expr::If {
                box test,
                box then_branch,
                box else_branch,
            } => Expr::If {
                id,
                test: box Expr::from_ast_inner(test, gen),
                then_branch: box Expr::from_ast_inner(then_branch, gen),
                else_branch: box Expr::from_ast_inner(else_branch, gen),
            },
            ast::Expr::Let { bindings, box body } => Expr::Let {
                id,
                bindings: bindings
                    .into_iter()
                    .map(|(var, val)| (var, Expr::from_ast_inner(val, gen)))
                    .collect(),

                body: box Expr::from_ast_inner(body, gen),
            },
            ast::Expr::Letrec { bindings, box body } => Expr::Letrec {
                id,
                bindings: bindings
                    .into_iter()
                    .map(|(var, val)| (var, Expr::from_ast_inner(val, gen)))
                    .collect(),
                body: box Expr::from_ast_inner(body, gen),
            },
            ast::Expr::Lambda { params, box body } => Expr::Lambda {
                id,
                params,
                body: box Expr::from_ast_inner(body, gen),
            },
            ast::Expr::App { box func, box arg } => Expr::App {
                id,
                func: box Expr::from_ast_inner(func, gen),
                arg: box Expr::from_ast_inner(arg, gen),
            },
        }
    }
}
