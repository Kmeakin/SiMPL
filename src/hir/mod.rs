pub use crate::types::ty::Type;
use crate::{syntax::ast, types::ty::TypeVarGen};
use derive_more::Display;
pub use simple_symbol::Symbol;
use std::str::FromStr;

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Lit {
        ty: Type,
        val: Lit,
    },
    Var {
        ty: Type,
        name: Symbol,
    },
    If {
        ty: Type,
        test: Box<Self>,
        then: Box<Self>,
        els: Box<Self>,
    },
    Let {
        ty: Type,
        binding: LetBinding,
        body: Box<Self>,
    },
    Letrec {
        ty: Type,
        bindings: Vec<LetBinding>,
        body: Box<Self>,
    },
    Lambda {
        ty: Type,
        param: Param,
        body: Box<Self>,
    },
    App {
        ty: Type,
        func: Box<Self>,
        arg: Box<Self>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub struct Param {
    pub ty: Type,
    pub name: Symbol,
    pub ann: Option<Type>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LetBinding {
    pub ty: Type,
    pub name: Symbol,
    pub ann: Option<Type>,
    pub val: Box<Expr>,
}

#[derive(Debug, Copy, Clone, PartialEq, Display)]
pub enum Lit {
    #[display(fmt = "{}", _0)]
    Bool(bool),
    #[display(fmt = "{}", _0)]
    Int(i64),
    #[display(fmt = "{}", _0)]
    Float(f64),
}

impl From<bool> for Lit {
    fn from(other: bool) -> Self {
        Self::Bool(other)
    }
}

impl From<i64> for Lit {
    fn from(other: i64) -> Self {
        Self::Int(other)
    }
}

impl From<f64> for Lit {
    fn from(other: f64) -> Self {
        Self::Float(other)
    }
}

impl From<ast::Lit> for Lit {
    fn from(other: ast::Lit) -> Self {
        match other {
            ast::Lit::Int(x) => Self::Int(x),
            ast::Lit::Float(x) => Self::Float(x),
            ast::Lit::Bool(x) => Self::Bool(x),
        }
    }
}

impl FromStr for Expr {
    type Err = String;
    fn from_str(src: &str) -> Result<Self, Self::Err> {
        // TODO: return a trait object instead of unwrapping
        let ast = crate::syntax::parse(src).unwrap();
        Ok(Self::from_ast(ast))
    }
}

impl Expr {
    /// `ast::Expr` -> `hir::Expr`
    /// Expands nested let/lambda, and attatches fresh type variables to every
    /// expression/binder
    pub fn from_ast(ast: ast::Expr) -> Self {
        let mut gen = TypeVarGen::new();
        Self::from_ast_inner(ast, &mut gen)
    }

    fn from_ast_inner(ast: ast::Expr, gen: &mut TypeVarGen) -> Self {
        match ast {
            ast::Expr::Lit { val } => Self::Lit {
                val: val.into(),
                ty: gen.next(),
            },
            ast::Expr::Var { name } => Self::Var {
                name,
                ty: gen.next(),
            },
            ast::Expr::Binop { .. } => todo!(),
            ast::Expr::If { test, then, els } => Self::If {
                ty: gen.next(),
                test: box Self::from_ast_inner(*test, gen),
                then: box Self::from_ast_inner(*then, gen),
                els: box Self::from_ast_inner(*els, gen),
            },
            ast::Expr::Let { bindings, body } => {
                let (binding, body) = expand_let(&bindings, *body);
                Self::Let {
                    ty: gen.next(),
                    binding: LetBinding {
                        ty: gen.next(),
                        name: binding.name,
                        ann: binding.ann,
                        val: box Self::from_ast_inner(*binding.val, gen),
                    },
                    body: box Self::from_ast_inner(body, gen),
                }
            }
            ast::Expr::Letrec { bindings, body } => Self::Letrec {
                ty: gen.next(),
                bindings: bindings
                    .into_iter()
                    .map(|binding| LetBinding {
                        ty: gen.next(),
                        name: binding.name,
                        ann: binding.ann,
                        val: box Self::from_ast_inner(*binding.val, gen),
                    })
                    .collect(),
                body: box Self::from_ast_inner(*body, gen),
            },
            ast::Expr::Lambda { params, body } => {
                let (param, body) = expand_lambda(&params, *body);
                Self::Lambda {
                    ty: gen.next(),
                    param: Param {
                        name: param.name,
                        ann: param.ann,
                        ty: gen.next(),
                    },
                    body: box Self::from_ast_inner(body, gen),
                }
            }
            ast::Expr::App { func, arg } => Self::App {
                ty: gen.next(),
                func: box Self::from_ast_inner(*func, gen),
                arg: box Self::from_ast_inner(*arg, gen),
            },
        }
    }

    pub fn ty(&self) -> Type {
        match self {
            Self::Lit { ty, .. }
            | Self::Var { ty, .. }
            | Self::If { ty, .. }
            | Self::Let { ty, .. }
            | Self::Letrec { ty, .. }
            | Self::Lambda { ty, .. }
            | Self::App { ty, .. } => ty.clone(),
        }
    }
}

fn expand_lambda(params: &[ast::Param], body: ast::Expr) -> (ast::Param, ast::Expr) {
    assert!(!params.is_empty());
    if params.len() == 1 {
        (params[0].clone(), body)
    } else {
        let param = &params[0];
        let (rest_param, rest_body) = expand_lambda(&params[1..], body);
        (
            param.clone(),
            ast::Expr::Lambda {
                params: vec![rest_param],
                body: box rest_body,
            },
        )
    }
}

fn expand_let(bindings: &[ast::LetBinding], body: ast::Expr) -> (ast::LetBinding, ast::Expr) {
    assert!(!bindings.is_empty());
    if bindings.len() == 1 {
        (bindings[0].clone(), body)
    } else {
        let binding = &bindings[0];
        let (rest_bindings, rest_body) = expand_let(&bindings[1..], body);
        (
            binding.clone(),
            ast::Expr::Let {
                bindings: vec![rest_bindings],
                body: box rest_body,
            },
        )
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::syntax::ast::{Expr, Lit};
    use simple_symbol::intern;

    #[test]
    fn test_expand_lambda() {
        let params = vec![
            ast::Param {
                name: intern("x"),
                ann: None,
            },
            ast::Param {
                name: intern("y"),
                ann: None,
            },
        ];
        let body = Expr::Lit { val: Lit::Int(0) };

        assert_eq!(
            expand_lambda(&params, body),
            (
                ast::Param {
                    name: intern("x"),
                    ann: None
                },
                Expr::Lambda {
                    params: vec![ast::Param {
                        name: intern("y"),
                        ann: None
                    }],
                    body: box Expr::Lit { val: Lit::Int(0) }
                }
            ),
        )
    }

    #[test]
    fn test_expand_let() {
        let bindings = vec![
            ast::LetBinding {
                name: intern("x"),
                val: box Expr::Lit { val: Lit::Int(1) },
                ann: None,
            },
            ast::LetBinding {
                name: intern("y"),
                val: box Expr::Lit { val: Lit::Int(2) },
                ann: None,
            },
        ];

        let body = Expr::Lit { val: Lit::Int(0) };

        assert_eq!(
            expand_let(&bindings, body),
            (
                ast::LetBinding {
                    name: intern("x"),
                    val: box Expr::Lit { val: Lit::Int(1) },
                    ann: None,
                },
                Expr::Let {
                    bindings: vec![ast::LetBinding {
                        name: intern("y"),
                        val: box Expr::Lit { val: Lit::Int(2) },
                        ann: None,
                    }],
                    body: box Expr::Lit { val: Lit::Int(0) }
                }
            ),
        )
    }
}
