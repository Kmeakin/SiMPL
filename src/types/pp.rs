use crate::hir::{Expr, LetBinding};
use pretty::RcDoc;

const INDENT: isize = 4;
const WIDTH: usize = 40;

fn binding_to_doc<'a>(binding: &'a LetBinding) -> RcDoc<'a> {
    RcDoc::nil()
        .append(binding.name.to_string())
        .append(RcDoc::space())
        .append(RcDoc::text("="))
        .append(RcDoc::space())
        .append(binding.val.to_doc())
}

impl Expr {
    pub fn to_doc(&self) -> RcDoc<()> {
        match self {
            Self::Lit { val, .. } => RcDoc::as_string(val),
            Self::Var { name, .. } => RcDoc::as_string(name),
            Self::If {
                test,
                then_branch,
                else_branch,
                ..
            } => RcDoc::text("if")
                .append(RcDoc::space())
                .append(test.to_doc())
                .append(RcDoc::line())
                .append(RcDoc::text("then"))
                .append(RcDoc::space())
                .append(then_branch.to_doc())
                .append(RcDoc::line())
                .append(RcDoc::text("else"))
                .append(RcDoc::space())
                .append(else_branch.to_doc())
                .nest(INDENT)
                .group(),
            Self::App { func, arg, .. } => RcDoc::text("(")
                .append(func.to_doc())
                .append(RcDoc::space())
                .append(arg.to_doc())
                .append(RcDoc::text(")"))
                .group(),
            Self::Lambda { param, body, .. } => RcDoc::text(r"\")
                .append(param.name.to_string())
                .append(RcDoc::space())
                .append(RcDoc::text("->"))
                .append(RcDoc::space())
                .append(body.to_doc())
                .group(),
            Self::Let { binding, body, .. } => RcDoc::text("let")
                .append(RcDoc::space())
                .append(binding_to_doc(binding))
                .append(RcDoc::space())
                .append(RcDoc::text("in"))
                .append(RcDoc::line())
                .append(body.to_doc())
                .nest(INDENT)
                .group(),
            Self::Letrec { bindings, body, .. } => {
                RcDoc::text("letrec").append(RcDoc::space()).append(
                    RcDoc::concat(bindings.iter().enumerate().map(|(i, b)| {
                        binding_to_doc(b)
                            .append(if i < bindings.len() - 1 {
                                RcDoc::text(",").append(RcDoc::hardline())
                            } else {
                                RcDoc::nil()
                            })
                            .nest(("letrec".len() + 1) as isize)
                            .group()
                    }))
                    .append(RcDoc::line())
                    .nest(-2 * INDENT)
                    .append(RcDoc::text("in"))
                    .nest(INDENT)
                    .append(RcDoc::line())
                    .append(body.to_doc())
                    .nest(INDENT)
                    .group(),
                )
            }
        }
    }

    pub fn pretty(&self) -> String {
        let mut v = Vec::new();
        self.to_doc().render(WIDTH, &mut v).unwrap();
        String::from_utf8(v).unwrap()
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use insta::assert_snapshot;
    use std::str::FromStr;

    #[track_caller]
    fn test_pp(src: &str) {
        let expr = Expr::from_str(src).unwrap();
        assert_snapshot!(expr.pretty());
    }

    #[test]
    fn pp_lit() {
        test_pp("123");
    }

    #[test]
    fn pp_var() {
        test_pp("add");
    }

    #[test]
    fn pp_if() {
        test_pp("if true then 1 else 0");
        test_pp("if (if false then true else true) then (if false then 1 else 0) else (if false then 420 else 69)");
    }

    #[test]
    fn pp_app() {
        test_pp("if not false then 1 else 0");
        test_pp("if not (is_zero (add 1 1)) then 50 else 100");
    }

    #[test]
    fn pp_lambda() {
        test_pp(r"if not false then \x -> x else \x -> not x");
        test_pp(r"if not false then \a, b -> a else \x, y -> y");
    }

    #[test]
    fn pp_let() {
        test_pp(r"let x = 5 in x");
        test_pp(r"let id = \x -> x, first = \a, b -> a in id not (first true 1)");
    }

    #[test]
    fn pp_letrec() {
        test_pp(r"letrec f = \x -> f x in f 0");
        test_pp(r"letrec f = \x -> x, g = \y -> y in f g");
        test_pp(r"letrec f1 = \a -> a, f2 = \b -> b, f3 = \c -> c, f4 = \d -> d in f1 f2 f3 f4");
    }
}
