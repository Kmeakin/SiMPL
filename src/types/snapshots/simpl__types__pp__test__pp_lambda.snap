---
source: src/types/pp.rs
expression: expr.pretty()
---
if (not false)
    then \x -> x
    else \x -> (not x)
